### PR TITLE
Refactor IntErrorKind to avoid "underflow" terminology

### PR DESCRIPTION
**Pull request by eopb**
This PR is a continuation of #76455 

# Changes

- `Overflow` renamed to `PosOverflow` and `Underflow` renamed to `NegOverflow` after discussion in #76455 
- Changed some of the parsing code to return `InvalidDigit` rather than `Empty` for strings "+" and "-". https://users.rust-lang.org/t/misleading-error-in-str-parse-for-int-types/49178
- Carry the problem `char` with the `InvalidDigit` variant.
- Necessary changes were made to the compiler as it depends on `int_error_matching`.
- Redid tests to match on specific errors.

r? @KodrAus

### CODE REVIEW

File: compiler/rustc_middle/src/middle/limits.rs [hunk 1]
```diff
@@ -48,10 +48,10 @@ fn update_limit(
                         .unwrap_or(attr.span);
 
                     let error_str = match e.kind() {
-                        IntErrorKind::Overflow => "`limit` is too large",
+                        IntErrorKind::PosOverflow => "`limit` is too large",
                         IntErrorKind::Empty => "`limit` must be a non-negative integer",
-                        IntErrorKind::InvalidDigit => "not a valid integer",
-                        IntErrorKind::Underflow => bug!("`limit` should never underflow"),
+                        IntErrorKind::InvalidDigit(_) => "not a valid integer",
+                        IntErrorKind::NegOverflow => bug!("`limit` should never underflow"),
```
Comment (KodrAus) [inline compiler/rustc_middle/src/middle/limits.rs:54]: If we're removing the _underflow_ terminology we should probably do it in the message here too.

File: library/core/src/num/error.rs [hunk 2]
```diff
@@ -77,51 +77,47 @@ pub struct ParseIntError {
...
     /// Among other causes, this variant will be constructed when parsing an empty string.
+    #[stable(feature = "int_error_matching", since = "1.47.0")]
     Empty,
-    /// Contains an invalid digit.
+    /// Contains an digit invalid in its context.
     ///
     /// Among other causes, this variant will be constructed when parsing a string that
-    /// contains a letter.
-    InvalidDigit,
+    /// contains a non-asci char.
```
Comment (KodrAus) [inline library/core/src/num/error.rs:98]:
```suggestion
    /// contains a non-ASCII char.
```

File: library/core/src/num/error.rs [hunk 3]
```diff
@@ -77,51 +77,47 @@ pub struct ParseIntError {
...
+    ///
+    /// This variant is also constructed when a `+` or `-` is misplaced within a sting
```
Comment (KodrAus) [inline library/core/src/num/error.rs:100]:
```suggestion
    /// This variant is also constructed when a `+` or `-` is misplaced within a string
```

File: library/core/src/num/error.rs [hunk 4]
```diff
@@ -77,51 +77,47 @@ pub struct ParseIntError {
...
 #[derive(Debug, Clone, PartialEq, Eq)]
 #[non_exhaustive]
 pub enum IntErrorKind {
     /// Value being parsed is empty.
     ///
     /// Among other causes, this variant will be constructed when parsing an empty string.
+    #[stable(feature = "int_error_matching", since = "1.47.0")]
     Empty,
-    /// Contains an invalid digit.
+    /// Contains an digit invalid in its context.
```
Comment (KodrAus) [inline library/core/src/num/error.rs:95]:
```suggestion
    /// Contains an invalid digit in its context.
```

File: library/core/src/num/error.rs [hunk 5]
```diff
@@ -77,51 +77,47 @@ pub struct ParseIntError {
...
     ///
     /// Among other causes, this variant will be constructed when parsing a string that
-    /// contains a letter.
-    InvalidDigit,
+    /// contains a non-asci char.
+    ///
+    /// This variant is also constructed when a `+` or `-` is misplaced within a sting
+    /// either on its own or in the middle of a number.
+    #[stable(feature = "int_error_matching", since = "1.47.0")]
+    InvalidDigit(#[stable(feature = "int_error_matching", since = "1.47.0")] char),
```
Comment (KodrAus) [inline library/core/src/num/error.rs:103]:
```suggestion
    InvalidDigit,
```

I think if we're going to keep the enum open here we should try limit it to simple variants. That would mean we don't know which `char` we encountered that caused us to return this variant, but in general I think just returning a `char` isn't quite enough for consumers to render meaningful errors. They'd probably also want to know the position in the string that digit appears at.

So I think removing the `char` from this variant will avoid a bit of noise in match arms without losing too much.

**Review by KodrAus — COMMENTED** [review]
Thanks for working on this @ethanboxx! I've left a few little comments.

Since we're making some changes to the shape of the `ParseIntError` I think we should avoid stabilizing it _just_ yet. Maybe give it a cycle or two with the new variants.

**Comment by ethanboxx** [comment]
@KodrAus Thank you for your review. This should be ready now.

**Comment by KodrAus** [comment]
Thanks @ethanboxx! This looks good to me.

@bors r+

**Pull request merged**